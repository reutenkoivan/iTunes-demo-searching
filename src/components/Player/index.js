import React, {Component} from 'react';
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import SkipPreviousIcon from '@material-ui/icons/SkipPrevious';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import Pause from "@material-ui/icons/Pause"
import SkipNextIcon from '@material-ui/icons/SkipNext';
import Typography from "@material-ui/core/Typography";

const styles = (theme) => ({
  appBar: {
    top: 'auto',
    bottom: 0,
  },
  title: {
    display: 'none',
    [theme.breakpoints.up('sm')]: {
      display: 'block',
    }
  },
  songInfo:{
    position: "absolute"
  },
  toolbar:{
    display: 'flex',
    alignItems: 'center',
  },
  controls: {
    display: 'flex',
    alignItems: 'center',
    margin: '0 auto',
  }
});

class Player extends Component {

  switchSoundState = () => {
    const typeOfPage = this.props.history.location.pathname.substring(1);
    const {favorites, songs} = this.props.track;
    const {favoritesSongs, searchingSongs, playerInfo} = this.props;

    if(playerInfo.artistName) {

      favorites.length
        ? this.props.addSongToPlayer(favorites[0], favoritesSongs)
        : this.props.addSongToPlayer(songs[0], searchingSongs);

    } else if(typeOfPage === "favorites" && favoritesSongs[0]){
      this.props.addSongToPlayer(favoritesSongs[0], favoritesSongs)
    } else if(!typeOfPage && searchingSongs[0])
      this.props.addSongToPlayer(searchingSongs[0], searchingSongs)
  };

  playNext = () => {
    let { playlist, next, artistName } = this.props.playerInfo;
    if(artistName && playlist.length !== 1)
      this.props.addSongToPlayer(playlist[next], playlist)
  };

  playPrevious = () => {
    let { playlist, previous, artistName } = this.props.playerInfo;
    if(artistName && playlist.length !== 1)
      this.props.addSongToPlayer(playlist[previous], playlist)
  };

  render(){
    const { classes, playerInfo} = this.props;
    return (
      <AppBar color="primary" className={classes.appBar}>
        <Toolbar className={classes.toolbar}>
          <div className={classes.songInfo}>
            <Typography name="trackName" className={classes.title} variant="h5" color="inherit" noWrap>
              {playerInfo.trackName}
            </Typography>
            <Typography name="artistName" className={classes.title} variant="h6" color="secondary" noWrap>
              {playerInfo.artistName}
            </Typography>
          </div>
          <div className={classes.controls}>
            <IconButton onClick={this.playPrevious} color="secondary" aria-label="Previous">
              <SkipPreviousIcon />
            </IconButton>
            <IconButton
              onClick={this.switchSoundState}
              aria-label={playerInfo.isPlaying
                ? "Pause"
                : "Play"
              }
            >
              {playerInfo.isPlaying
                ? <Pause color="secondary"/>
                : <PlayArrowIcon color="secondary"/>
              }
            </IconButton>
            <IconButton onClick={this.playNext} color="secondary" aria-label="Next">
              <SkipNextIcon />
            </IconButton>
          </div>
        </Toolbar>
      </AppBar>
    );
  }
}

export default withStyles(styles)(Player);
