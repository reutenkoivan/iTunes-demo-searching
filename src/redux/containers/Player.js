import { connect } from "react-redux";
import Player from "../../components/Player";
import { addSongSuccess } from "../actions";

const mapStateToProps = state => ({
  track: {
    songs: state.songs.filter(song => song.trackId === state.playerInfo.trackId),
    favorites: state.favorites.filter(song => song.trackId === state.playerInfo.trackId)
  },
  playerInfo: state.playerInfo,
  searchingSongs: state.songs,
  favoritesSongs: state.favorites
});

const mapDispatchToProps = dispatch => ({
  addSongToPlayer: (song, playlist) => dispatch(addSongSuccess(song, playlist))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Player);
