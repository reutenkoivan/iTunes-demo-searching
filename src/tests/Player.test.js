import React from 'react';
import Player from "../components/Player";
import { mount, shallow } from "enzyme";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Typography from "@material-ui/core/Typography";

configure({ adapter: new Adapter() });

const song = {
  artistName: "artistName",
  trackName: "trackName",
  previewUrl: "previewUrl",
  artworkUrl: "artworkUrl100",
  trackPrice: "trackPrice",
  trackId: "trackId",
  collectionName: "collectionName",
  artistViewUrl: "artistViewUrl",
  isFavorite: false
};
const playerInfo = {trackLink: "previewUrl",
  trackId: "trackId",
  artistName: "artistName",
  trackName: "trackName",
  playlist: [song, song],
  index: 1,
  next: 2,
  previous: 0
};
const track = {
  songs: [song],
  favorites: []
};

  it('renders without crashing', () => {
    const wrap = shallow(<Player/>);
    const newPlayerInfo = Object.assign({}, playerInfo);
    newPlayerInfo.isPlaying = false;
    wrap.setProps({ playerInfo: newPlayerInfo, track });

    expect(wrap).toMatchSnapshot()
  });


  // it('render with true', () => {
  //   const wrap = mount(<Player playerInfo={playerInfo}/>);
  //   const newPlayerInfo = Object.assign({}, playerInfo);
  //   newPlayerInfo.isPlaying = true;
  //   wrap.setProps({ playerInfo: newPlayerInfo, track });
  //   wrap.update();
  //   expect(wrap.props().playerInfo.isPlaying).toEqual(true);
  //   // expect(wrap.find(Typography).findWhere((f) => f.prop('variant') === 'h5')).toHaveLength(1)
  //   expect(wrap.find({ variant: 'h5'})).toHaveLength(2)
  //   expect(wrap).toMatchSnapshot()
  // });
