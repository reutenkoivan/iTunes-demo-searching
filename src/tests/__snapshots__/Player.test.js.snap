// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Player)
    playerInfo={
      Object {
        "artistName": "artistName",
        "index": 1,
        "isPlaying": true,
        "next": 2,
        "playlist": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
        "previous": 0,
        "trackId": "trackId",
        "trackLink": "previewUrl",
        "trackName": "trackName",
      }
    }
    track={
      Object {
        "favorites": Array [],
        "songs": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "appBar": "Player-appBar-1",
        "controls": "Player-controls-5",
        "songInfo": "Player-songInfo-3",
        "title": "Player-title-2",
        "toolbar": "Player-toolbar-4",
      },
      "playerInfo": Object {
        "artistName": "artistName",
        "index": 1,
        "isPlaying": true,
        "next": 2,
        "playlist": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
        "previous": 0,
        "trackId": "trackId",
        "trackLink": "previewUrl",
        "trackName": "trackName",
      },
      "track": Object {
        "favorites": Array [],
        "songs": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "appBar": "Player-appBar-1",
          "controls": "Player-controls-5",
          "songInfo": "Player-songInfo-3",
          "title": "Player-title-2",
          "toolbar": "Player-toolbar-4",
        },
        "playerInfo": Object {
          "artistName": "artistName",
          "index": 1,
          "isPlaying": true,
          "next": 2,
          "playlist": Array [
            Object {
              "artistName": "artistName",
              "artistViewUrl": "artistViewUrl",
              "artworkUrl": "artworkUrl100",
              "collectionName": "collectionName",
              "isFavorite": false,
              "previewUrl": "previewUrl",
              "trackId": "trackId",
              "trackName": "trackName",
              "trackPrice": "trackPrice",
            },
            Object {
              "artistName": "artistName",
              "artistViewUrl": "artistViewUrl",
              "artworkUrl": "artworkUrl100",
              "collectionName": "collectionName",
              "isFavorite": false,
              "previewUrl": "previewUrl",
              "trackId": "trackId",
              "trackName": "trackName",
              "trackPrice": "trackPrice",
            },
          ],
          "previous": 0,
          "trackId": "trackId",
          "trackLink": "previewUrl",
          "trackName": "trackName",
        },
        "track": Object {
          "favorites": Array [],
          "songs": Array [
            Object {
              "artistName": "artistName",
              "artistViewUrl": "artistViewUrl",
              "artworkUrl": "artworkUrl100",
              "collectionName": "collectionName",
              "isFavorite": false,
              "previewUrl": "previewUrl",
              "trackId": "trackId",
              "trackName": "trackName",
              "trackPrice": "trackPrice",
            },
          ],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Player)
    playerInfo={
      Object {
        "artistName": "artistName",
        "index": 1,
        "isPlaying": false,
        "next": 2,
        "playlist": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
        "previous": 0,
        "trackId": "trackId",
        "trackLink": "previewUrl",
        "trackName": "trackName",
      }
    }
    track={
      Object {
        "favorites": Array [],
        "songs": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "appBar": "Player-appBar-1",
        "controls": "Player-controls-5",
        "songInfo": "Player-songInfo-3",
        "title": "Player-title-2",
        "toolbar": "Player-toolbar-4",
      },
      "playerInfo": Object {
        "artistName": "artistName",
        "index": 1,
        "isPlaying": false,
        "next": 2,
        "playlist": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
        "previous": 0,
        "trackId": "trackId",
        "trackLink": "previewUrl",
        "trackName": "trackName",
      },
      "track": Object {
        "favorites": Array [],
        "songs": Array [
          Object {
            "artistName": "artistName",
            "artistViewUrl": "artistViewUrl",
            "artworkUrl": "artworkUrl100",
            "collectionName": "collectionName",
            "isFavorite": false,
            "previewUrl": "previewUrl",
            "trackId": "trackId",
            "trackName": "trackName",
            "trackPrice": "trackPrice",
          },
        ],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "appBar": "Player-appBar-1",
          "controls": "Player-controls-5",
          "songInfo": "Player-songInfo-3",
          "title": "Player-title-2",
          "toolbar": "Player-toolbar-4",
        },
        "playerInfo": Object {
          "artistName": "artistName",
          "index": 1,
          "isPlaying": false,
          "next": 2,
          "playlist": Array [
            Object {
              "artistName": "artistName",
              "artistViewUrl": "artistViewUrl",
              "artworkUrl": "artworkUrl100",
              "collectionName": "collectionName",
              "isFavorite": false,
              "previewUrl": "previewUrl",
              "trackId": "trackId",
              "trackName": "trackName",
              "trackPrice": "trackPrice",
            },
            Object {
              "artistName": "artistName",
              "artistViewUrl": "artistViewUrl",
              "artworkUrl": "artworkUrl100",
              "collectionName": "collectionName",
              "isFavorite": false,
              "previewUrl": "previewUrl",
              "trackId": "trackId",
              "trackName": "trackName",
              "trackPrice": "trackPrice",
            },
          ],
          "previous": 0,
          "trackId": "trackId",
          "trackLink": "previewUrl",
          "trackName": "trackName",
        },
        "track": Object {
          "favorites": Array [],
          "songs": Array [
            Object {
              "artistName": "artistName",
              "artistViewUrl": "artistViewUrl",
              "artworkUrl": "artworkUrl100",
              "collectionName": "collectionName",
              "isFavorite": false,
              "previewUrl": "previewUrl",
              "trackId": "trackId",
              "trackName": "trackName",
              "trackPrice": "trackPrice",
            },
          ],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
